core/src/operations/mod.rs

```diff
@@ -161,15 +161,19 @@ pub enum Operation {
     /// instruction.
     Clk = OPCODE_CLK,
 
-    /// Emits an event id (`u32` value) to the host.
-    ///
-    /// We interpret the event id as follows:
-    /// - 16 most significant bits identify the event source,
-    /// - 16 least significant bits identify the actual event.
-    ///
-    /// This operation does not change the state of the user stack. The event ID is read
-    /// from the top of the stack without consuming it. The event id must be manually
-    /// pushed and dropped.
+    /// Emits an event to the host.
+    ///
+    /// Semantics:
+    /// - Reads the event id from the top of the stack (as a `Felt`) without consuming it; the
+    ///   caller is responsible for pushing and later dropping the id.
+    /// - User-defined events are conventionally derived from strings via
+    ///   `hash_string_to_word(name)[0]` (Blake3-based) and may be emitted via immediate forms in
+    ///   assembly (`emit.event("...")` or `emit.CONST` where `CONST=event("...")`).
+    /// - System events are still identified by specific 32-bit codes; the VM attempts to interpret
+    ///   the stack `Felt` as `u32` to dispatch known system events, and otherwise forwards the
+    ///   event to the host.
```

Comment:
Is using `u32` IDs for system events temporary? (i.e., do we plan to remove this restriction in future PRs?)

Reply:
The system event IDs should probably be defined as `event("adv.has_map_key")` in which case it would be a full Felt. We can make this part of the effort to move events to the stdlib.

One small annoyance I wasn't able to fully figure out was how to create `const` IDs, since blake3 isn't const currently.

Reply:
Yep, we can do it in a follow-up.

> The system event IDs should probably be defined as `event("adv.has_map_key")` in which case it would be a full Felt.

I'm wondering if we could just assign constant values to them - e.g., `0`, `1`, `2` etc. I think with this approach a collision is equality unlikely as if we derived the values using BLAKE3.

Reply:
Linked in #1582